lib/bolt/application.rb

```diff
@@ -0,0 +1,612 @@
+                  "https://puppet.com/docs/puppet/latest/lang_classes.html and "\
+                  "https://puppet.com/docs/puppet/latest/lang_defined_types.html"
+        Bolt::Logger.warn("empty_manifest", message)
+      end
+
+      # Apply logging looks like plan logging
+      executor.publish_event(type: :plan_start, plan: nil)
+
+      with_benchmark do
+        pal.in_plan_compiler(executor, inventory, plugins.puppetdb_client) do |compiler|
+          compiler.call_function('apply_prep', targets)
+        end
+
+        pal.with_bolt_executor(executor, inventory, plugins.puppetdb_client) do
+          Puppet.lookup(:apply_executor).apply_ast(ast, targets, catch_errors: true, noop: noop)
+        end
+      end
+    end
+
+    # Run a command on a list of targets.
+    #
+    # @param command [String] The command.
+    # @param targets [Array[String]] The targets to run on.
+    # @param env_vars [Hash] Environment variables to set on the target.
+    # @return [Bolt::ResultSet]
+    #
+    def command_run(command, targets, env_vars: {})
+      targets = inventory.get_targets(targets)
+
+      with_benchmark do
+        executor.run_command(targets, command, env_vars: env_vars)
+      end
+    end
+
+    # Download a file from a list of targets to a directory on the controller.
+    #
+    # @param source [String] The path to the file on the targets.
+    # @param destination [String] The path to the directory on the controller.
+    # @param targets [Array[String]] The targets to run on.
+    # @return [Bolt::ResultSet]
+    #
+    def file_download(source, destination, targets)
+      destination = File.expand_path(destination, Dir.pwd)
+      targets     = inventory.get_targets(targets)
+
+      with_benchmark do
+        executor.download_file(targets, source, destination)
+      end
+    end
+
+    # Upload a file from the controller to a list of targets.
+    #
+    # @param source [String] The path to the file on the controller.
+    # @param destination [String] The destination path on the targets.
+    # @param targets [Array[String]] The targets to run on.
+    # @return [Bolt::ResultSet]
+    #
+    def file_upload(source, destination, targets)
+      future  = executor.future&.fetch('file_paths', false)
+      source  = find_file(source, future)
+      targets = inventory.get_targets(targets)
+
+      Bolt::Util.validate_file('source file', source)
+
+      with_benchmark do
+        executor.upload_file(targets, source, destination)
+      end
+    end
+
+    # Show groups in the inventory.
+    #
+    # @return [Hash]
+    #
+    def group_show
+      {
+        count: inventory.group_names.count,
+        groups: inventory.group_names.sort,
+        inventory: {
+          default: config.default_inventoryfile.to_s,
+          source: inventory.source
+        }
+      }
+    end
+
+    # TODO: Should this take an outputter or return data that the CLI
```

Comment:
I think returning data to be printed by the CLI makes more sense. It's what we do for things like `group show` and aligns more with each classes' role.

Reply:
Updated to return data for `group show` and error if given an unknown topic. Previously we showed the same output as `bolt guide`, but that gets a bit trickier to handle if `bolt guide <topic>` can return different data depending on if the topic is known or unknown. Erroring also seems to be in line with `show` commands that receive an unknown object.